Add unloadwallet RPC

promag opened the pull request:
This patch adds wallet unload feature via RPC. It also adds UI support for unloaded wallets.

promag committed:
rpc: Add unloadwallet RPC

promag commented:
Please only review last 3 commits.

promag committed:
doc: Add release notes for unloadwallet RPC

promag committed:
test: Add functional tests for unloadwallet RPC

promag commented:
Rebased, added release notes and functional tests.

promag commented:
If a wallet is loaded after being unloaded then a segfault is raised:
```sh
Process 10688 stopped
* thread #2, name = 'bitcoin-httpworker', stop reason = EXC_BAD_ACCESS (code=1, address=0x608)
    frame #0: 0x0000000100a0ad18 libdb_cxx-4.8.dylib`DbEnv::set_lg_dir(char const*) + 38
libdb_cxx-4.8.dylib`DbEnv::set_lg_dir:
->  0x100a0ad18 <+38>: callq  *0x608(%rdi)
    0x100a0ad1e <+44>: movl   %eax, %ebx
    0x100a0ad20 <+46>: testl  %ebx, %ebx
    0x100a0ad22 <+48>: je     0x100a0ad40               ; <+78>
Target 0: (bitcoind) stopped.
(lldb) bt
* thread #2, name = 'bitcoin-httpworker', stop reason = EXC_BAD_ACCESS (code=1, address=0x608)
  * frame #0: 0x0000000100a0ad18 libdb_cxx-4.8.dylib`DbEnv::set_lg_dir(char const*) + 38
    frame #1: 0x000000010026f0bf bitcoind`BerkeleyEnvironment::Open(this=0x0000000101900e68, retry=<unavailable>) at db.cpp:150 [opt]
    frame #2: 0x0000000100274a8a bitcoind`BerkeleyBatch::VerifyEnvironment(file_path=<unavailable>, errorStr="") at db.cpp:335 [opt]
    frame #3: 0x000000010030bd1f bitcoind`CWallet::Verify(wallet_file="�.\n\0p\0\0\x19\0\0\0\0\0\0\0\0\0���\x7f\0\0\x19\0\0\0\0\0\0\0p�.\n\0p\0\0\x012�w�, salvage_wallet=<unavailable>, error_string=<unavailable>, warning_string="") at wallet.cpp:4022 [opt]
    frame #4: 0x00000001002c1419 bitcoind`loadwallet(request=<unavailable>) at rpcwallet.cpp:3098 [opt]
    frame #5: 0x00000001001820ba bitcoind`CRPCTable::execute(this=<unavailable>, request=0x000070000a2eaab0) const at server.cpp:497 [opt]
    frame #6: 0x000000010001cc57 bitcoind`HTTPReq_JSONRPC(req=0x0000000101a7e1d0, (null)=<unavailable>) at httprpc.cpp:191 [opt]
    frame #7: 0x000000010002977b bitcoind`HTTPWorkItem::operator()() [inlined] std::__1::function<bool (HTTPRequest*, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)>::operator(this=<unavailable>, __arg=0x0000000101a7e1d0, __arg=<unavailable>)(HTTPRequest*, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) const at functional:1914 [opt]
    frame #8: 0x000000010002976e bitcoind`HTTPWorkItem::operator(this=<unavailable>)() at httpserver.cpp:54 [opt]
    frame #9: 0x000000010002b060 bitcoind`WorkQueue<HTTPClosure>::Run(this=0x0000000101801430) at httpserver.cpp:113 [opt]
    frame #10: 0x000000010002c03a bitcoind`void* std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, void (*)(WorkQueue<HTTPClosure>*), WorkQueue<HTTPClosure>*> >(void*) [inlined] decltype(std::__1::forward<void (*)(WorkQueue<HTTPClosure>*)>(fp)(std::__1::forward<WorkQueue<HTTPClosure>*>(fp0))) std::__1::__invoke<void (*)(WorkQueue<HTTPClosure>*), WorkQueue<HTTPClosure>*>(void (*&&)(WorkQueue<HTTPClosure>*), WorkQueue<HTTPClosure>*&&) at type_traits:4291 [opt]
    frame #11: 0x000000010002c032 bitcoind`void* std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, void (*)(WorkQueue<HTTPClosure>*), WorkQueue<HTTPClosure>*> >(void*) [inlined] void std::__1::__thread_execute<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, void (*)(WorkQueue<HTTPClosure>*), WorkQueue<HTTPClosure>*, 2ul>(std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, void (*)(WorkQueue<HTTPClosure>*), WorkQueue<HTTPClosure>*>&, std::__1::__tuple_indices<2ul>) at thread:336 [opt]
    frame #12: 0x000000010002c032 bitcoind`void* std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, void (*)(WorkQueue<HTTPClosure>*), WorkQueue<HTTPClosure>*> >(__vp=0x0000000101ad8af0) at thread:346 [opt]
    frame #13: 0x00007fff77cd66c1 libsystem_pthread.dylib`_pthread_body + 340
    frame #14: 0x00007fff77cd656d libsystem_pthread.dylib`_pthread_start + 377
    frame #15: 0x00007fff77cd5c5d libsystem_pthread.dylib`thread_start + 13
```

promag committed:
bugfix: Remove dangling wallet env instance

laanwj commented:
new warning:
```
/home/orion/projects/bitcoin/bitcoin/src/qt/bitcoingui.cpp:123:5: warning: field 'platformStyle' will be initialized after field 'm_wallet_selector_label' [-Wreorder]
    platformStyle(_platformStyle),
    ^
1 warning generated.
```

promag committed:
ui: Support wallets unloaded dynamically

ken2812221 commented:
The `wallet.dat` might corrupt by running the script:
```
src/bitcoind -daemon -regtest
sleep 3
src/bitcoin-cli -regtest createwallet test
src/bitcoin-cli -regtest unloadwallet test
num=0
while [ $num -lt 1000 ]; do
    src/bitcoin-cli -regtest loadwallet test
    src/bitcoin-cli -regtest unloadwallet test
    num=$(($num+1))
    echo $num
done

promag commented:
> The wallet.dat might corrupt by running the script

@ken2812221 corrupt how?

ken2812221 commented:
Oh, the wallet does not corrupt. It's a libevent error, but the message says that the wallet corrupts. It might because I open so many connections at a short period of time, restart bitcoind fix the problem. Does not related to this PR
```
2018-06-06T10:23:12Z  wallet                12033ms
2018-06-06T10:23:12Z setKeyPool.size() = 2000
2018-06-06T10:23:12Z mapWallet.size() = 0
2018-06-06T10:23:12Z mapAddressBook.size() = 0
2018-06-06T10:23:12Z libevent: Error from accept() call: Too many open files
2018-06-06T10:23:12Z BerkeleyEnvironment::Close: Error -30973 closing database environment: BDB0087 DB_RUNRECOVERY: Fatal error, run database recovery
2018-06-06T10:23:12Z Using BerkeleyDB version Berkeley DB 5.3.28: (September  9, 2013)
2018-06-06T10:23:12Z Using wallet wallet.dat
2018-06-06T10:23:12Z BerkeleyEnvironment::Open: LogDir=/home/user/.bitcoin/regtest/wallets/test/database ErrorFile=/home/user/.bitcoin/regtest/wallets/test/db.log
```

promag commented:
```
2018-06-06T10:23:12Z BerkeleyEnvironment::Close: Error -30973 closing database environment: BDB0087 DB_RUNRECOVERY: Fatal error, run database recovery
```
Is this related to the libevent error?

ken2812221 commented:
2018-06-06T10:23:12Z libevent: Error from accept() call: Too many open files

Probobly the wallet would fail to close after the libevent error. Restart the node fix the issue.

jnewbery commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3161,6 +3161,39 @@ UniValue createwallet(const JSONRPCRequest& request)
+    UnregisterValidationInterface(wallet.get());
+    wallet->BlockUntilSyncedToCurrentChain();
+    wallet->Flush(true);
+    wallet->Unload();
```
In commit *rpc: Add unloadwallet RPC*, this doesn't build (`CWallet.Unload()` is only added in *ui: Support wallets unloaded dynamically*)

promag replied:
Fixed.

jnewbery commented on test/functional/wallet_multiwallet.py:
```diff
@@ -234,5 +234,31 @@ def run_test(self):
+        assert_equal(self.nodes[0].listwallets(), [])
+
+        # Successfully load a previously unloaded wallet
+        self.nodes[0].loadwallet('w1')
```
Please assert that wallet **w1** appears in `listwallets` and is usable.

promag replied:
Done in 6efc7f4.

jnewbery reviewed (commented):
There seems to be a bug where if you run `loadwallet <wallet>` and then `unloadwallet <wallet>`, `<wallet>` isn't removed from the dropdown menu (if you call `unloadwallet <wallet>` on a wallet that was loaded at startup, it *is* removed.)

In general, the GUI doesn't seem to recognise unloading wallets that were loaded dynamically.

promag commented on src/wallet/wallet.h:
```diff
@@ -1097,6 +1097,13 @@ class CWallet final : public CCryptoKeyStore, public CValidationInterface
     void Flush(bool shutdown=false);
 
+    /**
+     * Wallet was unloaded.
```
Incorrect comment 😪

promag commented on src/wallet/wallet.h:
```diff
@@ -1101,6 +1101,8 @@ class CWallet final : public CCryptoKeyStore, public CValidationInterface
      * Address book entry changed.
      * @note called with lock cs_wallet held.
      */
+    boost::signals2::signal<void ()> Unload;
```
Yap.

promag commented:
@jnewbery thanks the review and testing. I believe I've addressed all your comments.

`wallet_model->moveToThread(thread());` was missing in the previous version which messed some `QObject::connect`.

Also, calling RPC `stop` either with bitcoin-qt or bitcoind caused a segfault, now fixed.

ACK on the unloadwallet usage calling without an argument defaults to the wallet referenced by the request URL?

promag commented:
```
GUI: QObject::connect: Cannot queue arguments of type 'QVector<int>'
(Make sure 'QVector<int>' is registered using qRegisterMetaType().)
```
@jonasschnelli this error was caused by incorrect `WalletModel`'s thread affinity, see above.

jnewbery commented on src/qt/walletmodel.h:
```diff
@@ -261,6 +262,9 @@ class WalletModel : public QObject
     void notifyWatchonlyChanged(bool fHaveWatchonly);
 
+    // Signal that wallet is about to be removed
+    void unload();
```
Is this `unload()` method doing anything?

jnewbery replied:
oops. Actually I think this declaration is required for `BitcoinApplication` to connect the `removeWallet` slot to the `unload` signal (https://github.com/bitcoin/bitcoin/pull/13111/files#diff-8c9d79ba40bf702f01685008550ac100R471). Is that right?

jnewbery reviewed (commented):
Looking very good. A few more comments inline

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3161,6 +3161,48 @@ UniValue createwallet(const JSONRPCRequest& request)
+    // Release the "main" shared pointer and prevent further notifications.
+    // Note that any attempt to load the same wallet would fail until the wallet
+    // is destroyed (see CheckUniqueFileid).
+    RemoveWallet(wallet);
```
Indeed, will fix.

promag replied:
Done.

promag commented on test/functional/wallet_multiwallet.py:
```diff
@@ -234,5 +234,33 @@ def run_test(self):
+        # Successfully unload the specified wallet name
+        self.nodes[0].unloadwallet("w1")
+        assert 'w1' not in self.nodes[0].listwallets()
+        w2.unloadwallet("w3")
```
So if the endpoint references a wallet then the RPC should have no arguments? If so LGTM.

promag replied:
Done.

promag commented:
@jnewbery thanks for the review, will address your comments. Also, I need to figure out the travis failure.

promag commented on src/qt/bitcoingui.cpp:
```diff
@@ -566,6 +567,19 @@ bool BitcoinGUI::addWallet(WalletModel *walletModel)
     return walletFrame->addWallet(walletModel);
 }
 
+bool BitcoinGUI::removeWallet(WalletModel* walletModel)
```
If you don't mind I would like to do it later. I have a couple of UI changes in the drawer.

promag replied:
@jnewbery done here.

promag committed:
rpc: Extract GetWalletNameFromJSONRPCRequest from GetWalletForJSONRPCRequest

promag committed:
test: Wallet methods are disabled when no wallet is loaded

achow101 commented:
Tested ACK 4eb7e12ba833c78201cb32a2f4a075dd5407396c

laanwj commented:
Code changes LGTM - utACK 9e816ff5cb151a50e6e57386b09cd42d540c650b

jnewbery commented on test/functional/wallet_multiwallet.py:
```diff
@@ -234,5 +234,34 @@ def run_test(self):
+        # Successfully unload the specified wallet name
+        self.nodes[0].unloadwallet("w1")
+        assert 'w1' not in self.nodes[0].listwallets()
+        # assert 'w3' not in self.nodes[0].listwallets()
```
Remove?

jnewbery commented on src/wallet/rpcwallet.cpp:
```diff
@@ -66,11 +75,6 @@ bool EnsureWalletIsAvailable(CWallet * const pwallet, bool avoidException)
     if (pwallet) return true;
     if (avoidException) return false;
     if (!HasWallets()) {
-        // Note: It isn't currently possible to trigger this error because
```
:kissing_smiling_eyes: :ok_hand:

jnewbery reviewed (commented):
As mentioned in IRC, `WalletModel` needs the `unload()` declaration readded.

One other nit in the test.

promag commented:
@jnewbery thanks for the review!

jonasschnelli commented:
Issue QT:
1. Start with a single (default) wallet
2. Create or load a wallet (`[default wallet]` and `[new_wallet]` should be there in the dropdown).
3. When `[default wallet]` is selected, call `unloadwallet`
4.  **`[default wallet]` is still there in the RPC dropdown**
5.  **`[default wallet]` is still in the main window dropdown**
6.  **Selecting `[default wallet]` looks like one have selected `[default wallet]` but acctually `new_wallet` is selected**

jonasschnelli commented:
Tested ACK 0496cfbf7b28bf65eb5726e5b119ff75d56e5df8 (after squash).

jonasschnelli commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3161,6 +3165,54 @@ UniValue createwallet(const JSONRPCRequest& request)
+    }
+
+    std::string wallet_name;
+    if (GetWalletNameFromJSONRPCRequest(request, wallet_name)) {
```
nit: it's unclear from the calls documentation that you can't unload `walletB` by calling `walletA`-endpoint with `unloadwallet  walletB`.

promag replied:
@jonasschnelli improved documentation.

PierreRochard commented:
Segfault with QT:

1. Start with a single (default) wallet
2. Create or load a wallet ([default wallet] and [new_wallet] should be there in the dropdown).
3. When [new_wallet] is selected in the dropdown, call unloadwallet

<pre>
Process 87183 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=EXC_I386_GPFLT)
    frame #0: 0x00000001000b3edc bitcoin-qt`SendCoinsDialog::updateSmartFeeLabel(this=0x0000000124329810) at sendcoinsdialog.cpp:697 [opt]
   694 	    coin_control.m_feerate.reset(); // Explicitly use only fee estimation rate for smart fee labels
   695 	    int returned_target;
   696 	    FeeReason reason;
-> 697 	    CFeeRate feeRate = CFeeRate(model->wallet().getMinimumFee(1000, coin_control, &returned_target, &reason));
   698 	
   699 	    ui->labelSmartFee->setText(BitcoinUnits::formatWithUnit(model->getOptionsModel()->getDisplayUnit(), feeRate.GetFeePerK()) + "/kB");
   700 	
</pre>

![image](https://user-images.githubusercontent.com/6299681/41604637-b69608f4-73ad-11e8-98ee-fb5054e3cbe9.png)

I'm still debugging it but wanted to bring it up and see if others can reproduce.

promag commented:
@PierreRochard thanks for testing. I'm able to reproduce it. After your steps (which don't harm), call `generate 1`. This is due to the missing `delete`s in https://github.com/bitcoin/bitcoin/blob/3f398d7a17f136cd4a67998406ca41a124ae2966/src/qt/walletview.cpp#L87-L90
Will fix.

promag committed:
bugfix: Delete walletView in WalletFrame::removeWallet

promag commented:
@PierreRochard actually `WalletView` was leaking. Added a separate commit since the problem already exists in master.

jnewbery commented:
Tested ACK fe65bdec237776dbe094339509dfd2e63329a832

@jonasschnelli and @PierreRochard issues are fixed.

PierreRochard commented:
@promag, thanks for the explanation, I really like this pull request. 
I think we should handle the case where `GetWalletNameFromJSONRPCRequest` returns false and there is no `wallet_name` name provided. Otherwise the endpoint chokes on `wallet_name = request.params[0].get_str();` and provides an unexpected error to the user

![image](https://user-images.githubusercontent.com/6299681/41674180-411ff1d0-748d-11e8-93aa-c9564b2f87dd.png)

Maybe the issue is that URI should be populated after unload/load, it currently isn't:

![image](https://user-images.githubusercontent.com/6299681/41674316-a83e01ae-748d-11e8-8dc5-a6cf084914a8.png)

promag commented:
@PierreRochard it can be improved in a follow up.

PierreRochard commented:
That's fair, tested ACK fe65bdec237776dbe094339509dfd2e63329a832

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -79,6 +79,15 @@ std::shared_ptr<CWallet> GetWallet(const std::string& name)
+static void ReleaseWallet(CWallet* wallet)
+{
+    LogPrintf("Releasing wallet %s\n", wallet->GetName());
+    wallet->BlockUntilSyncedToCurrentChain();
```
What's the reason for this call? It seems very fragile because BlockUntilSyncedToCurrentChain may try to schedule a callback, but wallets currently get deleted after the scheduler is deleted:

https://github.com/bitcoin/bitcoin/blob/baf3a3ab0c63b512b37d9f753768c1f020369088/src/init.cpp#L297-L299

I was seeing this cause a crash due to a subtle change in behavior in one of my prs, which took me a long time to debug.

promag replied:
The initial idea was to avoid dangling pointers to the wallet in the validation queue. I agree that once in the deleter no more pointers to the wallet should exist. Do you suggest removing this call? Which PR are you referring?

ryanofsky replied:
> The initial idea was to avoid dangling pointers to the wallet in the validation queue.

@promag I think it would be better to just call UnregisterValidationInterface in this case. That looks like it would remove any wallet references from the notification code without needing to block wallet deletion on the scheduler thread or run additional notification handling code.

>  I agree that once in the deleter no more pointers to the wallet should exist. Do you suggest removing this call? Which PR are you referring?

Yesterday the problem I saw was in #10973. It slightly changed the way `m_last_block_processed` got set during rescans leading to a new `SyncWithValidationInterfaceQueue()` call in `SyncWithValidationInterfaceQueue` that doesn't happen in master. The `SyncWithValidationInterfaceQueue` call would fail with an assert error due to signal scheduler being deleted before the wallet in the init code shown above. 

Today this is causing another problem in #11625. In this case, there seems to be a race condition there where the wallet sometimes gets deleted in a qt event thread after `UnloadBlockIndex` gets called in another thread.
